src/Ontology/Address.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Hash.h"
+#include "Address.h"
+#include "HexCoding.h"
+
+#include <stdexcept>
+
+#include <TrezorCrypto/sha2.h>
+#include <TrezorCrypto/base58.h>
+#include <TrezorCrypto/ripemd160.h>
+
+using namespace TW::Ontology;
+
+Address::Address(const PublicKey &publicKey) {
+    auto hash = std::array<uint8_t, Hash::sha256Size>();
+    std::vector<uint8_t> builder(publicKey.bytes);
+    builder.insert(builder.begin(), 0x21);
+    builder.push_back(0xAC);
+    auto data = toScriptHash(builder);
+    std::copy(data.begin(), data.end(), zero.begin());
+}
+
+
+Address::Address(const std::string &b58Address) {
+    size_t dataSz = 25;
+    std::array<uint8_t, 25> data{};
+    b58tobin(data.data(), &dataSz, b58Address.data());
+    if (data[0] != version) {
+        throw std::runtime_error("Invalid address version.");
+    }
+    std::vector<uint8_t> addressWithVer(size + 1);
+    std::copy(data.begin(), data.begin() + size + 2, addressWithVer.begin());
+    auto checksum = Hash::sha256(Hash::sha256(addressWithVer));
+    auto count = 1;
+    for (auto v:checksum) {
+        if (count == 5) {
+            break;
+        }
+        if (v != *(data.begin() + size + count)) {
+            throw std::runtime_error("Invalid base58 encode address.");
+        }
+        count++;
+    }
+    std::copy(addressWithVer.begin() + 1, addressWithVer.end(), zero.begin());
+}
+
+Address::Address(const std::vector<uint8_t> &data) {
+    if (!Address::isValid(data)) {
+        throw std::runtime_error("Invalid vector");
+    }
+    std::copy(data.begin(), data.end(), zero.begin());
+}
+
+std::vector<uint8_t> Address::toScriptHash(std::vector<uint8_t> &data) {
+    return Hash::ripemd(Hash::sha256(data));
+}
+
+bool Address::isValid(const std::string &string) {
+    const auto data = parse_hex(string);
+    return Address::isValid(data);
+}
+
+std::string Address::hexString() const {
```

Comment:
Do we really need hex string address format? Where to use it?

Reply:
The smart contract address is in the form of hex string & The wallet account address is in the form of base58 encode string.

Reply:
https://github.com/TrustWallet/wallet-core/pull/103/commits/8eac3fbeb8a318e3a7139983fe6012f9ec3bc28f